ref/test/utilityfunctions/isEqual.f90

```diff
@@ -0,0 +1,83 @@
+module isequalmodule
+  
+ contains 
+
+  !------------------------------------------------------------
+  !       isequal 
+  !
+  !       Function which compares equality of two real*8 floats
+  !       within a given tolerance value.
+  !------------------------------------------------------------
+
+  logical function isequal(a, b, tolerance)
+    implicit none
+
+    real*8, intent(in)   :: a, b, tolerance
+    real*8               :: absA, absB, diff
+
+    absA = abs(a)
+    absB = abs(b)
+    diff = abs(a - b)
+
+    if (a == b) then
+        isequal = .true.
+    else if ( a == 0 .or. b == 0 .or. (absA + absB < TINY(a))) then
```

Comment:
Is there a case where (absA + absB) < TINY(a) is true?  I think not.  My understanding of tiny(a) is that it is the smallest positive floating-point number of type a.

Reply:
@LyndStringer - This algorithm is based on the one given [here](https://floating-point-gui.de/errors/comparison/). I think probably what you are pointing to is a subtle misunderstanding of what that line is doing.  In the original Java code that line is:
```
else if (a == 0 || b == 0 || (absA + absB < Float.MIN_NORMAL))
```
What that line is trying to capture is the situation where floating point numbers get denormalized such that they are smaller than the smallest **normal** value a float can represent.  This can and does [happen](https://software.intel.com/content/www/us/en/develop/articles/fortran-numbers-smaller-than-tiny.html), though there are consequences.  Java's `Float.MIN_NORMAL` does appear to be equivalent to Fortran's `TINY`. function. `TINY` is the smallest **normal** value that is representable; it is not the smallest representable value.

Reply:
You are right, TINY is the smallest positive non-zero floating point number, not a denormal.  

[Lynd.Stringer@hfe11 ~]$ cat test1.f90
implicit none
real a4
double precision a8
print *,"4-byte real tiny=",tiny(a4)
print *,"4-byte real tiny/2=",tiny(a4)/2
print *,"8-byte real= tiny",tiny(a8) 
print *,"8-byte real tiny/2=",tiny(a8)/2
end

[Lynd.Stringer@hfe11 ~]$ ifort test1.f90;./a.out
 4-byte real tiny=  1.1754944E-38
 4-byte real tiny/2=  5.8774718E-39
 8-byte real= tiny  2.225073858507201E-308
 8-byte real tiny/2=  1.112536929253601E-308

So let's say we have two 4-byte fp denormals, a = 5.5E-39 and b =5.5E-39.  They are not zero.  
Their sum is less than TINY(a).
The next line is: "isequal = diff < (tolerance * TINY(a))" where tolerance*TINY(a) is likely zero.  
So isequal is set to diff<0.0 or FALSE.